lib/client.js

```diff
@@ -661,31 +661,50 @@ MatrixClient.prototype.joinRoom = function(roomIdOrAlias, opts, callback) {
     if (utils.isFunction(opts)) {
         throw new Error("Expected 'opts' object, got function.");
     }
-    opts = opts || {
-        syncRoom: true
-    };
+    opts = opts || {};
+    if (opts.syncRoom === undefined) { opts.syncRoom = true; }
 
     var room = this.getRoom(roomIdOrAlias);
     if (room && room.hasMembershipState(this.credentials.userId, "join")) {
         return q(room);
     }
-    var path = utils.encodeUri("/join/$roomid", { $roomid: roomIdOrAlias});
+
+    var sign_promise = q();
+
+    if (opts.inviteSignUrl) {
+        sign_promise = this._http.requestOtherUrl(
+            undefined, 'POST',
+            opts.inviteSignUrl, { mxid: this.credentials.userId }
+        );
+    }
+
     var defer = q.defer();
+
     var self = this;
-    this._http.authedRequest(undefined, "POST", path, undefined, {}).then(
-    function(res) {
-        var roomId = res.room_id;
-        var syncApi = new SyncApi(self);
-        var room = syncApi.createRoom(roomId);
-        if (opts.syncRoom) {
-            // v2 will do this for us
-            // return syncApi.syncRoom(room);
+    sign_promise.done(function(signed_invite_object) {
+        var data = {};
+        if (signed_invite_object) {
+            data.third_party_signed = signed_invite_object;
         }
-        return q(room);
-    }, function(err) {
-        _reject(callback, defer, err);
-    }).done(function(room) {
-        _resolve(callback, defer, room);
+
+        var path = utils.encodeUri("/join/$roomid", { $roomid: roomIdOrAlias});
+        self._http.authedRequest(undefined, "POST", path, undefined, data).then(
+        function(res) {
```

Comment:
rather than nesting the promise handlers and ending up with a pyramid of doom, how about:

```
sign_promise.then(function(signed_invite_object) {
    ...
    return self._http.authedRequest(undefined, "POST", path, undefined, data);
}).then(function(res) {
    var roomId = res.room_id;
    ...
    return q(room);
}).done(function(room) {
    _resolve(callback, defer, room);
}, function(err) {
    _reject(callback, defer, err);
});
```